File: ruby/lib/datagrout_conduit/transport/ws.rb
# frozen_string_literal: true

require "websocket/driver"
require "socket"
require "openssl"
require "uri"
require "thread"
require "json"
require "securerandom"
require "base64"
require "timeout"

module DatagroutConduit
  module Transport
    # WebSocket transport for datagrout-jsonrpc.v1.
    #
    # Manages a single wss:// connection with concurrent JSON-RPC request
    # multiplexing and server-push subscriptions. Uses a background thread
    # for frame reading; callers block on Thread::Queue for responses.
    #
    # Usage:
    #   ws = DatagroutConduit::Transport::Ws.new(
    #     url: "wss://gateway.datagrout.ai/servers/<uuid>/ws",
    #     auth: { bearer: "token" }
    #   )
    #   ws.connect
    #   result = ws.send_request("tools/list")
    #   sub    = ws.subscribe("agents.my-agent.events")
    #   event  = sub.recv
    #   ws.unsubscribe(sub)
    #   ws.disconnect
    class Ws
      SUBPROTOCOL = "datagrout-jsonrpc.v1"

      # ── Subscription ─────────────────────────────────────────────────────────

      # Per-subscription event stream delivered via a thread-safe Queue.
      # Call recv to block until the next event, or iterate with each.
      class Subscription
        attr_reader :sub_id, :topic

        def initialize(sub_id, topic)
          @sub_id = sub_id
          @topic  = topic
          @queue  = ::Thread::Queue.new
          @closed = false
        end

        # Block until the next event arrives.
        # Returns nil and raises StopIteration on close when iterating via each.
        # @param timeout [Numeric, nil] optional timeout in seconds; returns nil on expiry
        def recv(timeout: nil)
          event =
            if timeout
              Timeout.timeout(timeout) { @queue.pop }
            else
              @queue.pop
            end

          raise StopIteration if event.nil?

          event
        rescue Timeout::Error
          nil
        end

        # Iterate over events until the subscription is closed.
        def each(&block)
          loop do
            event = @queue.pop
            break if event.nil?
            block.call(event)
          end
        end

        include Enumerable

        def closed?
          @closed
        end

        # @api private
        def _enqueue(event)
          @queue.push(event) unless @closed
        end

        # @api private
        def _close
          return if @closed

          @closed = true
          @queue.push(nil)
        end
      end

      # Value object for a push notification delivered to a subscription.
      SubscriptionEvent = Struct.new(:subscription, :event, :data, keyword_init: true)

      # ── Construction ─────────────────────────────────────────────────────────

      def initialize(url:, auth: {}, identity: nil)
        @url      = url
        @auth     = normalize_auth(auth)
        @identity = identity
        @mutex    = Mutex.new
        @write_mutex = Mutex.new

        @pending           = {}  # id => RequestFuture
        @pending_subscribe = {}  # id => { topic:, future: }
        @subscriptions     = {}  # sub_id => [Subscription, ...]
        @next_id           = 0

        @io          = nil
        @driver      = nil
        @read_thread = nil
        @connected   = false
      end

      # ── Public API ────────────────────────────────────────────────────────────

      # Establish the WebSocket connection.
      # Blocks until the server-side handshake completes (up to 10 s).
      def connect
        uri = URI.parse(@url)
        @io = open_socket(uri)

        adapter = SocketAdapter.new(@url, @io)
        @driver = WebSocket::Driver.client(adapter, protocols: [SUBPROTOCOL])

        build_upgrade_headers.each { |k, v| @driver.set_header(k, v) }

        handshake_q = ::Thread::Queue.new

        @driver.on(:open)    { handshake_q.push(nil) unless @connected }
        @driver.on(:message) { |e| handle_message(e.data) }
        @driver.on(:close)   { handle_disconnect }
        @driver.on(:error)   { |e| handshake_q.push(e.message) unless @connected }

        @driver.start
        @read_thread = Thread.new { read_loop }
        @read_thread.abort_on_exception = false
        @read_thread.name = "conduit-ws-reader"

        err = Timeout.timeout(10) { handshake_q.pop }
        raise ConnectionError, "WebSocket handshake failed: #{err}" if err

        @connected = true
        self
      rescue Timeout::Error
        cleanup_socket
        raise ConnectionError, "WebSocket connection timed out"
      rescue ConnectionError
        raise
      rescue => e
        cleanup_socket
        raise ConnectionError, "WebSocket connect error: #{e.message}"
      end

      # Close the connection and fail all pending requests.
      def disconnect
        @mutex.synchronize { @connected = false }
        fail_all_pending(:disconnected)
        cleanup_socket
        self
      end

      def connected?
        @connected
      end

      # Send a JSON-RPC request and block until the response arrives.
      # Pass id: nil to fire a notification (no response expected).
      # Returns the result value, or raises McpError on RPC-level error.
      def send_request(method, params = nil, id: :auto)
        ensure_connected!

        # id: nil means fire-and-forget notification (no id field, no response wait)
        if id.nil?
          frame = { "jsonrpc" => "2.0", "method" => method }
          frame["params"] = params if params
          write_frame(frame)
          return { "result" => {} }
        end

        req_id = mint_id
        future = RequestFuture.new
        @mutex.synchronize { @pending[req_id] = future }

        write_frame(build_request(req_id, method, params))

        result, value = future.wait
        if result == :ok
          { "result" => value }
        else
          raise McpError.new(code: -1, message: value.to_s, data: nil)
        end
      end

      # Subscribe to a dotted-namespace topic.
      # Returns a Subscription that delivers events via recv / each.
      def subscribe(topic)
        ensure_connected!

        req_id = mint_id
        future = RequestFuture.new
        @mutex.synchronize { @pending_subscribe[req_id] = { topic: topic, future: future } }

        write_frame(build_request(req_id, "subscribe", { "topic" => topic }))

        result, value = future.wait
        if result == :ok
          sub_id = value.is_a?(Hash) ? (value["subscription"] || req_id) : req_id
          sub = Subscription.new(sub_id, topic)
          @mutex.synchronize do
            @subscriptions[sub_id] ||= []
            @subscriptions[sub_id] << sub
          end
          sub
        else
          raise McpError.new(code: -1, message: value.to_s, data: nil)
        end
      end

      # Cancel a push subscription locally and notify the server.
      # Accepts a Subscription object or a subscription ID string.
      def unsubscribe(subscription)
        sub_id = subscription.is_a?(Subscription) ? subscription.sub_id : subscription.to_s

        subs = @mutex.synchronize { @subscriptions.delete(sub_id) || [] }
        subs.each(&:_close)

        if @connected
          req_id = mint_id
          write_frame(build_request(req_id, "unsubscribe", { "subscription" => sub_id }))
        end

        :ok
      end

      private

      # ── Socket adapter for websocket-driver ──────────────────────────────────

      class SocketAdapter
        attr_reader :url

        def initialize(url, io)
          @url = url
          @io  = io
        end

        def write(data)
          @io.write(data)
          data.bytesize
        end
      end

      # ── Request future ────────────────────────────────────────────────────────

      class RequestFuture
        def initialize
          @queue = ::Thread::Queue.new
        end

        def wait(timeout: 30)
          Timeout.timeout(timeout) { @queue.pop }
        rescue Timeout::Error
          [:error, "Request timed out after #{timeout}s"]
        end

        def resolve(value)
          @queue.push([:ok, value])
        end

        def reject(reason)
          @queue.push([:error, reason])
        end
      end

      # ── Socket creation ───────────────────────────────────────────────────────

      def open_socket(uri)
        host = uri.host
        port = uri.port || (uri.scheme == "wss" ? 443 : 80)

        tcp = TCPSocket.new(host, port)
        tcp.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)

        if uri.scheme == "wss"
          ctx = build_ssl_context
          ssl = OpenSSL::SSL::SSLSocket.new(tcp, ctx)
          ssl.hostname = host
          ssl.sync_close = true
          ssl.connect
          ssl
        else
          tcp
        end
      end

      def build_ssl_context
        ctx = OpenSSL::SSL::SSLContext.new
        ctx.set_params(verify_mode: OpenSSL::SSL::VERIFY_PEER)

        if @identity
          ctx.cert = @identity.openssl_cert
          ctx.key  = @identity.openssl_key
          if @identity.ca_pem
            store = OpenSSL::X509::Store.new
            store.add_cert(@identity.openssl_ca)
            ctx.cert_store = store
          end
        end

        ctx
      end

      # ── Upgrade headers ───────────────────────────────────────────────────────

      def build_upgrade_headers
        headers = {}

        case @auth[:type]
        when :bearer
          headers["Authorization"] = "Bearer #{@auth[:token]}"
        when :api_key
          headers["X-API-Key"] = @auth[:key]
        when :basic
          encoded = Base64.strict_encode64("#{@auth[:username]}:#{@auth[:password]}")
          headers["Authorization"] = "Basic #{encoded}"
        when :oauth
          token = @auth[:provider].get_token
          headers["Authorization"] = "Bearer #{token}"
        end

        headers
      end

      # ── Read loop ─────────────────────────────────────────────────────────────

      def read_loop
        loop do
          data = @io.readpartial(4096)
          @driver.parse(data)
        rescue EOFError, IOError, Errno::ECONNRESET, Errno::EPIPE
          handle_disconnect
          break
        rescue StandardError
          handle_disconnect
          break
        end
      end

      # ── Message routing ───────────────────────────────────────────────────────

      def handle_message(raw)
        msg = JSON.parse(raw)

        str_id = msg["id"]&.to_s

        if str_id && (entry = @mutex.synchronize { @pending_subscribe.delete(str_id) })
          future = entry[:future]
          if msg["error"]
            future.reject(msg.dig("error", "message") || "Subscribe failed")
          else
            future.resolve(msg["result"] || {})
          end

        elsif str_id && (future = @mutex.synchronize { @pending.delete(str_id) })
          if msg["error"]
            future.reject(msg.dig("error", "message") || "RPC error")
          else
            future.resolve(msg["result"])
          end

        elsif msg["method"] == "notification"
          route_notification(msg["params"] || {})
        end
      rescue JSON::ParserError
        # Silently discard malformed frames
      end

      def route_notification(params)
        sub_id = params["subscription"]
        return unless sub_id.is_a?(String)

        subs = @mutex.synchronize { @subscriptions[sub_id] }
        return unless subs

        event = SubscriptionEvent.new(
          subscription: sub_id,
          event: params["event"] || "",
          data: params["data"]
        )

        subs.each { |sub| sub._enqueue(event) }
      end

      def handle_disconnect
        was_connected = @mutex.synchronize do
          old = @connected
          @connected = false
          old
        end

        fail_all_pending(:disconnected) if was_connected
      end

      def fail_all_pending(reason)
        pending, pending_sub, subs = @mutex.synchronize do
          p  = @pending.dup
          ps = @pending_subscribe.dup
          s  = @subscriptions.dup
          @pending.clear
          @pending_subscribe.clear
          @subscriptions.clear
          [p, ps, s]
        end

        pending.each_value     { |f| f.reject(reason) }
        pending_sub.each_value { |entry| entry[:future].reject(reason) }
        subs.each_value        { |list| list.each(&:_close) }
      end

      def cleanup_socket
        @write_mutex.synchronize do
          @driver = nil
        end
        @read_thread&.kill
        @read_thread = nil
        @io&.close rescue nil
        @io = nil
      end

      # ── Helpers ───────────────────────────────────────────────────────────────

      def ensure_connected!
        raise NotInitializedError, "WebSocket not connected. Call connect() first." unless @connected
      end

      def mint_id
        @mutex.synchronize do
          @next_id += 1
          "ws-#{@next_id}"
        end
      end

      def build_request(id, method, params)
        body = { "jsonrpc" => "2.0", "id" => id, "method" => method }
        body["params"] = params if params
        body
      end

      def write_frame(data)
        json = JSON.generate(data)
        @write_mutex.synchronize { @driver&.text(json) }
      end

      def normalize_auth(auth)
        return { type: :none } if auth.nil? || auth.empty?

        auth = auth.transform_keys(&:to_sym) if auth.is_a?(Hash)

        if auth[:bearer]
          { type: :bearer, token: auth[:bearer] }
        elsif auth[:api_key]
          { type: :api_key, key: auth[:api_key] }
        elsif auth[:basic]
          { type: :basic, username: auth[:basic][:username], password: auth[:basic][:password] }
        elsif auth[:oauth] || auth[:provider]
          { type: :oauth, provider: auth[:oauth] || auth[:provider] }
        else
          { type: :none }
        end
      end
    end
  end
end

File: elixir/lib/datagrout_conduit/transport/ws.ex
defmodule DatagroutConduit.Transport.Ws do
  @moduledoc """
  WebSocket transport for `datagrout-jsonrpc.v1`.

  Manages a single `wss://` connection with concurrent JSON-RPC request
  multiplexing and server-push subscriptions.  Concurrent requests are
  correlated by JSON-RPC `id` with no head-of-line blocking.

  ## Usage

      {:ok, ws} = DatagroutConduit.Transport.Ws.start_link(
        url: "wss://gateway.datagrout.ai/servers/<uuid>/ws",
        auth: {:bearer, "token"}
      )

      {:ok, result} = DatagroutConduit.Transport.Ws.send_request(ws, "tools/list")

  ## Push subscriptions

  Subscribe to dotted-namespace topics and receive events as messages in the
  calling process's mailbox:

      {:ok, sub_id} = DatagroutConduit.Transport.Ws.subscribe(ws, "agents.my-agent-id.events")

      receive do
        {:subscription_event, ^sub_id, event} ->
          IO.inspect(event)
      end

      :ok = DatagroutConduit.Transport.Ws.unsubscribe(ws, sub_id)

  ## Wire protocol

  - Subprotocol: `datagrout-jsonrpc.v1`
  - Connect URL: `wss://<gateway>/servers/<uuid>/ws`
  - Frame format: JSON-RPC 2.0, text frames only
  - Auth: `Authorization: Bearer <token>` in the upgrade headers
  """

  use GenServer

  require Logger

  alias DatagroutConduit.Transport.Ws.Conn

  @subprotocol "datagrout-jsonrpc.v1"

  @doc """
  Milliseconds between client-initiated WS ping frames.

  Many load balancers and reverse proxies (nginx, AWS ALB) close idle WS
  connections after 60-120 seconds; pinging every 25 seconds keeps the
  connection alive well within the tightest common timeout window.  Mirrors
  `PING_INTERVAL` in the Rust reference implementation.
  """
  @ping_interval_ms 25_000
  def ping_interval_ms, do: @ping_interval_ms

  # ── State ──────────────────────────────────────────────────────────────────

  defstruct [
    :conn_pid,
    pending: %{},
    pending_subscribe: %{},
    # sub_id => [subscriber_pid]
    subscriptions: %{},
    next_id: 0,
    # ms between ping frames; overridable via :ping_interval_ms init opt
    ping_interval_ms: @ping_interval_ms,
    # bumped each time a ping frame is sent (exposed via pings_sent/1 for tests)
    pings_sent: 0
  ]

  @doc "Returns how many ping frames this transport has sent. For tests."
  @spec pings_sent(GenServer.server()) :: non_neg_integer()
  def pings_sent(pid), do: GenServer.call(pid, :pings_sent)

  # ── Public API ─────────────────────────────────────────────────────────────

  @doc "Start a supervised WebSocket transport process."
  @spec start_link(keyword()) :: GenServer.on_start()
  def start_link(opts) do
    GenServer.start_link(__MODULE__, opts)
  end

  @doc """
  Send a JSON-RPC request over the WebSocket and wait for the response.

  Returns `{:ok, result}` or `{:error, reason}`.
  """
  @spec send_request(GenServer.server(), String.t(), map() | nil, timeout()) ::
          {:ok, term()} | {:error, term()}
  def send_request(pid, method, params \\ nil, timeout \\ 30_000) do
    GenServer.call(pid, {:send_request, method, params}, timeout)
  end

  @doc """
  Subscribe to a dotted-namespace push topic.

  Events are delivered as `{:subscription_event, subscription_id, event}` messages
  to the calling process's mailbox, where `event` is a map with `:event` and
  `:data` keys.

  Returns `{:ok, subscription_id}` on success.
  """
  @spec subscribe(GenServer.server(), String.t(), timeout()) ::
          {:ok, String.t()} | {:error, term()}
  def subscribe(pid, topic, timeout \\ 10_000) do
    GenServer.call(pid, {:subscribe, topic, self()}, timeout)
  end

  @doc """
  Cancel a server-side push subscription.

  Removes the subscription locally and notifies the server.
  Returns `:ok`.
  """
  @spec unsubscribe(GenServer.server(), String.t(), timeout()) :: :ok | {:error, term()}
  def unsubscribe(pid, subscription_id, timeout \\ 10_000) do
    GenServer.call(pid, {:unsubscribe, subscription_id}, timeout)
  end

  # ── GenServer init ─────────────────────────────────────────────────────────

  # Accept a pre-built %Ws{} struct directly — used in tests to inject state
  # without going through the real WebSocket connection setup.
  @impl true
  def init(%__MODULE__{} = state), do: {:ok, state}

  def init(opts) do
    url = Keyword.fetch!(opts, :url)
    auth = Keyword.get(opts, :auth)
    identity = Keyword.get(opts, :identity)
    ping_interval_ms = Keyword.get(opts, :ping_interval_ms, @ping_interval_ms)

    ws_url = to_ws_url(url)
    headers = build_headers(auth)

    conn_opts = [
      headers: headers,
      identity: identity,
      parent: self()
    ]

    case Conn.start_link(ws_url, conn_opts) do
      {:ok, conn_pid} ->
        state = %__MODULE__{conn_pid: conn_pid, ping_interval_ms: ping_interval_ms}
        schedule_ping(state)
        {:ok, state}

      {:error, reason} ->
        {:stop, reason}
    end
  end

  # Schedule the next ping tick.  No-op when ping_interval_ms <= 0 so tests
  # can disable pinging entirely by passing 0.
  defp schedule_ping(%__MODULE__{ping_interval_ms: interval}) when interval > 0 do
    Process.send_after(self(), :ping_tick, interval)
  end

  defp schedule_ping(_), do: nil

  # ── GenServer callbacks ────────────────────────────────────────────────────

  @impl true
  def handle_call({:send_request, method, params}, from, state) do
    {id, state} = next_id(state)
    frame = build_request(id, method, params)

    case send_frame(state.conn_pid, frame) do
      :ok ->
        pending = Map.put(state.pending, id, from)
        {:noreply, %{state | pending: pending}}

      {:error, reason} ->
        {:reply, {:error, reason}, state}
    end
  end

  def handle_call({:subscribe, topic, caller_pid}, from, state) do
    {id, state} = next_id(state)
    frame = build_request(id, "subscribe", %{"topic" => topic})

    case send_frame(state.conn_pid, frame) do
      :ok ->
        pending = Map.put(state.pending_subscribe, id, {topic, caller_pid, from})
        {:noreply, %{state | pending_subscribe: pending}}

      {:error, reason} ->
        {:reply, {:error, reason}, state}
    end
  end

  def handle_call({:unsubscribe, sub_id}, _from, state) do
    # Remove local subscription immediately.
    state = %{state | subscriptions: Map.delete(state.subscriptions, sub_id)}

    {id, state} = next_id(state)
    frame = build_request(id, "unsubscribe", %{"subscription" => sub_id})

    case send_frame(state.conn_pid, frame) do
      :ok ->
        # Best-effort — don't wait for the ack.
        {:reply, :ok, state}

      {:error, _reason} ->
        {:reply, :ok, state}
    end
  end

  def handle_call(:pings_sent, _from, state) do
    {:reply, state.pings_sent, state}
  end

  # ── Incoming frame routing ─────────────────────────────────────────────────

  @impl true
  def handle_info({:ws_frame, raw}, state) do
    state =
      case Jason.decode(raw) do
        {:ok, msg} -> handle_message(msg, state)
        {:error, _} -> state
      end

    {:noreply, state}
  end

  # Periodic ping tick — send a WS ping frame to keep the connection alive
  # past idle-timeout-happy load balancers (nginx, AWS ALB) and reschedule
  # the next tick.  See @ping_interval_ms doc for rationale.
  #
  # Wraps the send in try/catch because WebSockex.send_frame `exit`s rather
  # than returning when the Conn process has died — without this, every
  # disconnect would crash the Ws GenServer along with it.
  def handle_info(:ping_tick, state) do
    state =
      case safe_send_ping(state.conn_pid) do
        :ok -> %{state | pings_sent: state.pings_sent + 1}
        _ -> state
      end

    schedule_ping(state)
    {:noreply, state}
  end

  def handle_info(:ws_disconnected, state) do
    Logger.warning("[Ws] WebSocket disconnected — failing pending requests")
    reason = :disconnected

    for {_id, from} <- state.pending do
      GenServer.reply(from, {:error, reason})
    end

    for {_id, {_topic, _caller, from}} <- state.pending_subscribe do
      GenServer.reply(from, {:error, reason})
    end

    {:noreply, %{state | pending: %{}, pending_subscribe: %{}, subscriptions: %{}}}
  end

  def handle_info(_other, state), do: {:noreply, state}

  # ── Message routing ────────────────────────────────────────────────────────

  defp handle_message(%{"id" => nil} = msg, state),
    do: handle_message(Map.delete(msg, "id"), state)

  defp handle_message(%{"method" => "notification", "params" => params}, state)
       when not is_map_key(state, :id) do
    route_notification(params, state)
    state
  end

  defp handle_message(msg, state) when not is_map_key(msg, "id") do
    if msg["method"] == "notification" do
      route_notification(msg["params"] || %{}, state)
    end

    state
  end

  defp handle_message(%{"id" => id} = msg, state) do
    str_id = to_string(id)

    cond do
      Map.has_key?(state.pending_subscribe, str_id) ->
        {_topic, caller_pid, from} = Map.fetch!(state.pending_subscribe, str_id)
        pending_subscribe = Map.delete(state.pending_subscribe, str_id)

        if err = msg["error"] do
          GenServer.reply(from, {:error, err["message"] || "Subscribe failed"})
          %{state | pending_subscribe: pending_subscribe}
        else
          result = msg["result"] || %{}
          sub_id = result["subscription"] || str_id
          # Register the caller as a subscriber.
          subs = Map.update(state.subscriptions, sub_id, [caller_pid], &[caller_pid | &1])
          GenServer.reply(from, {:ok, sub_id})
          %{state | pending_subscribe: pending_subscribe, subscriptions: subs}
        end

      Map.has_key?(state.pending, str_id) ->
        from = Map.fetch!(state.pending, str_id)
        pending = Map.delete(state.pending, str_id)

        if err = msg["error"] do
          GenServer.reply(from, {:error, err["message"] || "RPC error"})
        else
          GenServer.reply(from, {:ok, msg["result"]})
        end

        %{state | pending: pending}

      true ->
        Logger.debug("[Ws] response for unknown id #{inspect(str_id)}")
        state
    end
  end

  defp route_notification(%{"subscription" => sub_id} = params, state) when is_binary(sub_id) do
    event = %{
      event: params["event"] || "",
      data: params["data"]
    }

    case Map.get(state.subscriptions, sub_id) do
      nil ->
        :ok

      pids ->
        for pid <- pids, Process.alive?(pid) do
          send(pid, {:subscription_event, sub_id, event})
        end
    end
  end

  defp route_notification(_params, _state), do: :ok

  # ── Internal helpers ───────────────────────────────────────────────────────

  defp next_id(%__MODULE__{next_id: n} = state) do
    {"ws-#{n + 1}", %{state | next_id: n + 1}}
  end

  defp build_request(id, method, params) do
    base = %{"jsonrpc" => "2.0", "id" => id, "method" => method}
    if params, do: Map.put(base, "params", params), else: base
  end

  defp send_frame(conn_pid, frame) do
    case Jason.encode(frame) do
      {:ok, json} -> Conn.send_frame(conn_pid, {:text, json})
      {:error, reason} -> {:error, reason}
    end
  end

  # Wraps `Conn.send_frame/2` with a try/catch so the Ws GenServer survives
  # a vanished Conn process — `WebSockex.send_frame/2` exits the caller when
  # the target client has gone away, which would otherwise take down this
  # GenServer alongside the dead connection.
  defp safe_send_ping(conn_pid) do
    try do
      Conn.send_frame(conn_pid, {:ping, ""})
    catch
      :exit, _ -> {:error, :conn_gone}
      _kind, _err -> {:error, :send_failed}
    end
  end

  defp to_ws_url(url) do
    url
    |> String.replace_prefix("https://", "wss://")
    |> String.replace_prefix("http://", "ws://")
  end

  defp build_headers(nil), do: [{"sec-websocket-protocol", @subprotocol}]

  defp build_headers({:bearer, token}),
    do: [{"authorization", "Bearer #{token}"}, {"sec-websocket-protocol", @subprotocol}]

  defp build_headers({:api_key, key}),
    do: [{"x-api-key", key}, {"sec-websocket-protocol", @subprotocol}]

  defp build_headers({:basic, user, pass}) do
    encoded = Base.encode64("#{user}:#{pass}")
    [{"authorization", "Basic #{encoded}"}, {"sec-websocket-protocol", @subprotocol}]
  end

  defp build_headers({:oauth, provider_pid}) do
    token = GenServer.call(provider_pid, :get_token)
    [{"authorization", "Bearer #{token}"}, {"sec-websocket-protocol", @subprotocol}]
  end
end
